pkg/skaffold/build/scheduler.go

```diff
@@ -35,45 +35,58 @@ type ArtifactBuilder func(ctx context.Context, out io.Writer, artifact *latest.A
 
 // For testing
 var (
-	buffSize      = bufferedLinesPerArtifact
-	runInSequence = InSequence
+	buffSize = bufferedLinesPerArtifact
 )
 
-// InParallel builds a list of artifacts in parallel but prints the logs in sequential order.
-func InParallel(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact, buildArtifact ArtifactBuilder, concurrency int) ([]Artifact, error) {
-	if len(artifacts) == 0 {
-		return nil, nil
-	}
+type artifactWithDeps struct {
+	artifact              *latest.Artifact
+	artifactChan          chan interface{}
+	requiredArtifactChans []chan interface{}
+}
 
-	if len(artifacts) == 1 || concurrency == 1 {
-		return runInSequence(ctx, out, tags, artifacts, buildArtifact)
+func (a artifactWithDeps) markComplete() {
+	// closing channel notifies all listeners waiting for this build to complete
+	close(a.artifactChan)
+}
+func (a artifactWithDeps) waitForDependencies(ctx context.Context) {
+	for _, dep := range a.requiredArtifactChans {
+		// wait for dependency to complete build
+		select {
+		case <-ctx.Done():
```

Comment:
do we need to catch `ctx.Cancelled()` here as well? or is that handled somewhere further down/up?

Reply:
I'm not sure I follow. `context` here has already been setup with cancellation so when `cancel` gets called in the main go routine it'll close this done channel.
From docs `WithCancel arranges for Done to be closed when cancel is called;`

Reply:
I get it know.
```
case <-ctx.Done():
    if ctx.Err() == ctx.Cancelled():
      ...
```

I'll add that. Thanks!